detox/src/devices/drivers/android/emulator/FreeEmulatorFinder.test.js

```diff
@@ -1,94 +1,32 @@
 
-  let mockDeviceRegistry;
   let uut;
   beforeEach(() => {
-    jest.mock('../../../../utils/logger');
-
-    const DeviceRegistry = jest.genMockFromModule('../../../DeviceRegistry');
-    mockDeviceRegistry = new DeviceRegistry();
-    mockDeviceFree();
-
-    const FreeEmulatorFinder = require('./FreeEmulatorFinder');
-    uut = new FreeEmulatorFinder(mockAdb, mockDeviceRegistry, mockAvdName);
+    uut = new FreeEmulatorFinder(undefined, undefined);
   });
 
-  it('should create an adb devices helper', async () => {
-    expect(AdbDevicesHelperClass).toHaveBeenCalledWith(mockAdb);
+  it('should match when is an emulator and avdName matches', async () => {
+    const candidate = createDevice(avdName);
+    expect(await uut.isDeviceMatching(candidate, avdName)).toBe(true);
   });
 
-  it('should pass a custom matcher func onto adb-devices helper for finding a free emulator', async () => {
-    await uut.findFreeDevice();
-    expect(adbDevicesHelperObj().lookupDevice).toHaveBeenCalledWith(expect.any(Function));
+  it('should not match when not an emulator', async () => {
+    const candidate = createDevice(avdName, 'device');
+    expect(await uut.isDeviceMatching(candidate, avdName)).toBe(false);
   });
 
-  it('should return the matched emulator', async () => {
-    const adbName = 'mock-adb-name';
-    adbDevicesHelperObj().lookupDevice.mockReturnValue(adbName);
-
-    const deviceName = await uut.findFreeDevice();
-    expect(deviceName).toEqual(adbName);
+  it('should not match when avdName does not match', async () => {
+    const candidate = createDevice(avdName);
+    expect(await uut.isDeviceMatching(candidate, 'wrongAvdName')).toBe(false);
   });
 
-  describe('Device lookup matcher function', () => {
-    it('should restrict matching to emulators', async () => {
-      const emuDevice = anEmulatorDevice();
-      const nonEmuDevice = aNonEmuDevice();
-
-      await uut.findFreeDevice();
-
-      const matcherFn = resolveMatcherFn();
-      expect(await matcherFn(emuDevice)).toEqual(true);
-      expect(await matcherFn(nonEmuDevice)).toEqual(false);
-    });
-
-    it('should restrict to free (unoccupied) devices', async () => {
-      const device = anEmulatorDevice();
-      mockDeviceBusy();
-
-      await uut.findFreeDevice();
-
-      const matcherFn = resolveMatcherFn();
-      expect(await matcherFn(device)).toEqual(false);
-      expect(mockDeviceRegistry.isDeviceBusy).toHaveBeenCalledWith('mock-name');
-    });
-
-    it('should restrict to the AVD name in question', async () => {
-      const device = {
-        ...anEmulatorDevice(),
-        queryName: jest.fn().mockResolvedValue('other ' + mockAvdName),
-      };
-
-      await uut.findFreeDevice();
-
-      const matcherFn = resolveMatcherFn();
-      expect(await matcherFn(device)).toEqual(false);
-    });
-
-    const resolveMatcherFn = () => adbDevicesHelperObj().lookupDevice.mock.calls[0][0];
+  const createDevice = (avdName, type = 'emulator') => ({
```

Comment:
I suggest moving out of `detox/src/devices/drivers/android/exec/ADB.js` those two classes:

* DeviceHandle
* EmulatorHandle

either to one file (e.g.., `DeviceHandle.js`), or two appropriate ones, let's say inside the folder `detox/src/devices/drivers/android/tools`.

Then you will be just able to use them in both of your tests, for `FreeDeviceFinder` and `FreeEmulatorFinder`.

Emulator: 

```js
const candidate = new EmulatorHandle('emulator-5555\tdevice');
jest.spyOn(candidate, 'queryName').mockImplementation(async () => 'mockAvdName');
expect(await uut.isDeviceMatching(candidate, 'wrongAvdName')).toBe(false);
```

Genymotion:

```js
const candidate = new DeviceHandle('192.168.60.101:6666\tdevice');
```

Attached device:

```js
const candidate = new DeviceHandle('MOCK_SERIAL\tdevice');
```

Incidentally, if you take a look at the implementation, you'll see we check for `candidate.type === 'emulator'` first not only for fun, but also because non-emulator device handles simply do not have this method, `async queryName()`, so it is not legit to have it always in your test doubles.

I believe the aforementioned steps will bring more cohesion into unit tests codebase because the reader will understand what those `candidate`s are in reality.

Reply:
P. S. Since you use candidates in 3 unit tests, it might make sense to create some `__mocks__/handles.js` helper file that exports one handle of each kind, like those mentioned above.

Reply:
Done